### ISSUE TITLE
[New Check] BadHighwayLinkCheck

### ISSUE THREAD

**Issue by nelkapelmenka** [osmlab/atlas-checks#598]
**Description**
Flagging the highway links that are labeled improperly or are too long (over 1000 meters).

**Requirements**
- Type way
- Has one of the following highway tags: ["motorway_link", "trunk_link", "primary_link", "secondary_link", "tertiary_link"]
- Case 1: Highway link too long: over 4 nodes AND over 1000 m in length
- Case 2: Highway link tag doesn't correspond to the higher priority highway on either side of the link.

**Use Cases**

Example:

*Case 1: [Highway too long.]*
 - A highway with a link tag that is over 4 nodes and 1000 meters long
 - *TP: [Way: 265460906](https://www.openstreetmap.org/way/265460906)*
![image](https://user-images.githubusercontent.com/55819977/128742726-978a9828-55b8-48cc-933a-fcd368f2af11.png)
![image](https://user-images.githubusercontent.com/55819977/128743012-9ce90129-a941-4d06-94a7-c692ffaeac83.png)

*Case 2: [Highway link named inappropriately.]*
- A highway link is connected to a highway that is higher priority than the link. In this example, a tertiary_link is connecting a trunk highway. Trunk is higher priority than tertiary, therefore the way needs to have a highway=trunk_link tag.
- *TP: [Way: 41716786](https://www.openstreetmap.org/way/417167686)*
![image](https://user-images.githubusercontent.com/55819977/128743305-8794157f-1231-41b5-b638-ecadd275b11b.png)

**Supported regions**
No Restrictions

**References**
[OSMOSE Wiki](https://wiki.openstreetmap.org/wiki/Osmose/issues#1110)
[OSMOSE Backend](https://github.com/osm-fr/osmose-backend/blob/c6c0e8a4598495c3b21c308081dbda92be0d1717/analysers/analyser_osmosis_highway_link.py)

**Comment by Bentleysb** [comment]
@nelkapelmenka 
This looks to be doing mostly the same thing as [RoadLinkCheck](https://github.com/osmlab/atlas-checks/blob/dev/docs/checks/roadLinkCheck.md). I think this would make sense as an enhancement for for that check, updating the default minimum to 1km, and adding the higher priority highway logic to the existing highway connection logic.

**Comment by nelkapelmenka** [comment]
@Bentleysb 
I have looked into RoadLinkCheck. You are right, it seems to be a quite similar check, in fact, checking the length is exactly the same, just different max variable. However, checking for the connections of the _links, it seems to be doing the same thing but following different principles.
RoadLinkCheck first check if the link is long enough and if it is it flags it as is and doesn't check connections. However, for my check I was working under the assumption that links should usually be less than 1 km, but it should be checked if it is a check. Hence, I am checking for both the length and the connections.
The next discrepancy is that this check only looks if any connections are of equal classification as the link which would cause a few issues for me.
First issue example, is the fact that the function .isOfEqualClassification() is true if the _link has a connection to another _link. I.e., [Way: 435208174](https://www.openstreetmap.org/way/435208174), the way is tertiary_link and it is on one end connected to a different "tertiary_link". This second _link isn't the actual proper highway connection, it is just a different _link that happens to be sharing the end node. While it is not true for this example, this could be an issue that a way isn't being flagged if the real connecting highways are of different classification.
Second issue example is the classification itself. In the [Wiki page](https://wiki.openstreetmap.org/wiki/Key:highway) it says "The link roads (sliproads/ramps) leading to/from a X road from/to a X road or lower class highway," where X is a type of highway based on X_link. Therefore, I want to not only check if one of the connections is equal classification but also if the other connection isn't of higher priority. As in, I would flag a primary_link if it is connecting primary and trunk highways, since trunk is of higher priority than primary.
Additionally, I have come across a _link that only had connections one side (oddly), which made me think of another edge case scenario, where links don't have an enter or an exit. This isn't ordinary, but could potentially be existent in some countries. I was planning (after talking to Kyle today) on adding another flag for these cases for the editors to confirm if this is actually a link, or if it has been misnamed.
**At this point, I am not sure if it makes sense to add all these to the RoadLinkCheck, or to create a separate check altogether.**

**Comment by Bentleysb** [comment]
It sounds to me like your logic is an improvement on what is already in RoadLinkCheck. So I would say just go ahead and modify road link to use your logic for ramp connections. If the one sided link turn out to be a real issue then feel free to add that as well.

**Comment by nelkapelmenka** [comment]
@Bentleysb sounds good, thank you for looking over this!

**Comment by nelkapelmenka** [comment]
@Bentleysb I have also added a couple more edge cases that I have found while working on this check.

**Comment by atiannicelli** [comment]
@nelkapelmenka are you still working on this issue?

**Comment by nelkapelmenka** [comment]
@atiannicelli no, I hit a lot of obstacles with this improvement that resulted in a significantly high FP rate (I believe it was about 30%). [This branch](https://github.com/nelkapelmenka/atlas-checks/tree/BadHighwayLinkCheck) is my latest experimental version.
